### ISSUE TITLE
Garden does not recover after a machine restart

### ISSUE THREAD

**Issue by gcapizzi** [cloudfoundry/garden-runc-release#195]
# TL;DR

Garden running in `containerd` mode with a sufficient number of containers will fail to recover after a machine restart. This is due to `containerd` taking a long time trying to re-adopt all the shims of the previous running containers.

# Details

After a machine restart, all running containers are gone, but they are still marked as _running_ in the `containerd` store. `containerd` then tries to reconnect to their shims, which are currently stored in `/run/containerd`, a `tmpfs` mount which is empty after a restart. These attempts fail with `ENOENT`, and due to a bug introduced in containerd/containerd#3802 lead to a 100 seconds wait. This will look like this in `containerd.stderr.log` (notice the 100 seconds difference in timestamps):

```
time="2021-06-21T11:26:03.615639875Z" level=warning msg="cleaning up after shim disconnected" id=0387f783-e92b-4445-6b7a-1594 namespace=garden
time="2021-06-21T11:26:03.615718903Z" level=info msg="cleaning up dead shim"
time="2021-06-21T11:26:04.324284541Z" level=warning msg="cleanup warnings time=\"2021-06-21T11:26:03Z\" level=info msg=\"starting signal loop\" namespace=garden pid=6971\n"
time="2021-06-21T11:27:44.434598858Z" level=warning msg="cleaning up after shim disconnected" id=043068d4-07a9-4b1b-70c3-d376 namespace=garden
time="2021-06-21T11:27:44.434644145Z" level=info msg="cleaning up dead shim"
```

With a sufficient number of containers, `containerd` will take a very long time to start up and probably trigger other timeouts (e.g. in BOSH/Monit).

More details, including codebase references, in containerd/containerd#5597. A fix to the `ENOENT` bug has already been submitted in containerd/containerd#5648.

# Possible solutions

[This comments](https://github.com/containerd/containerd/issues/5597#issuecomment-862912003) suggests that the `containerd` `state` should be stored in `tmpfs`, so that it gets cleared on restart. This should work: on restart containers will be terminated *and* state will be cleared, so no reconnection attempts should occur.

The [default value](https://github.com/containerd/containerd/blob/main/docs/man/containerd-config.toml.5.md) for the `state` config parameter is `/run/containerd`, but in `garden-runc-release` [we override it](https://github.com/cloudfoundry/garden-runc-release/blob/cb0328adf00dbcdf8c01e4391645909c8fc59bfa/jobs/garden/templates/config/containerd.toml.erb#L2) to `/var/vcap/data/containerd/state`, which is not in `tmpfs`. We believe moving `state` to a `tmpfs` mount might solve the problem. The [best location](https://bosh.io/docs/vm-config/#storage) to use in BOSH seems to be `/var/vcap/sys/run`.

We are not sure about `root`, another location used to store state, which is *not* in `tmpfs` by default (`/var/lib/containerd`). In case moving `state` is not enough, we should try moving `root` too.

If moving `state` works, we should make sure that we support upgrades transparently, by moving the existing folder from `/var/vcap/data/containerd/state` to the new `tmpfs` location (`/var/vcap/sys/run/containerd/state`?) on the first run.

If moving `state` (and potentially `root`) to a `tmpfs` doesn't work, we will have to wait for containerd/containerd#5648 to be merged and released.

**Comment by kieron-dev** [comment]
Simulate `state` being stored in a tmpfs by removing the state directory as well as `/run/containerd`:

1. Create containers alice and bob
2. `rm -rf /run/containerd /var/vcap/data/containerd/state`
3. `monit restart garden`

This time there is no 'cleaning up after shim disconnected' in the containerd logs. Containerd restarts in less than 0.01s. Monit reports garden running first time.

There is one discrepancy. Containerd reports containers alice and bob as still existing, as seen in the output of curling localhost:7777/containers: `{"Handles":["alice","bob"]}`. And in this simulation the container processes do still exist. But containerd has lost track of the state stored in the state directory and cannot contact the shims, so trying to run a process in one of those containers returns a 'task for container XXX not found' error.

**Comment by kieron-dev** [comment]
Simulate both `state` and `root` being stored in a tmpfs by removing both along with `/run/containerd`:

1. Create containers alice and bob
2. `rm -rf /run/containerd/ /var/vcap/data/containerd/state/ /var/vcap/data/containerd/root/`
3. `monit restart garden`

Results are similar to before. Containerd comes back up in around the same time. Garden starts successfully and monit shows it as running. However, containerd does not consider the containers as running.

This might be a cleaner scenario, as garden will not report to diego that the containers are running. They are in this simulation, but in a reboot they would not be. So containerd's view of the world would match reality.

**Comment by kieron-dev** [comment]
To test this properly we need to reboot a VM. Ideally there would be a way to do this via the bosh CLI, but there isn't. `bosh stop` is the equivalent of running `monit stop all` on the VM. `bosh stop --hard` deletes the VM, meaning we lose the ephemeral disk containing the problematic containerd state.

I had though doing a `reboot` from the VM didn't work. But you just need to wait long enough for the VM to come back up and for bosh to reconnect to its agent. This takes several minutes. Then you can `bosh ssh` back onto the VM and watch containerd cleaning up the dead shims with 100s intervals.

So `reboot` is the way forward.

**Comment by kieron-dev** [comment]
The above commit changes the containerd state directory to use an ephemeral location. We have concerns over the rootless code's ability to use this location, and will watch CI for any problems.

**Comment by kieron-dev** [comment]
The rootless tests are ok. The work was abandoned, and is only tested in runc mode. So this doesn't affect it.
